### PR TITLE
(core) modularize route configuration, convert to TS

### PR DESCRIPTION
**Pull request by anotherchrisberry**
We've already started on this with the various states.js files, but this takes it a little further, breaking the states out for instances, clusters, tasks, projects, etc.

The main state configuration now goes through state.provider.ts, which is pretty simple: you can add a new root state, or add a rewrite rule. All application states are configured via the application.state.provider.ts provider.

`$stateParams` are no longer generic (something @danielpeach started some time ago with the execution details state params), so we're getting good type checking there, and will get good typing elsewhere and don't need to use bracket syntax to access state parameters.

The last change is building an actual Application model for standalone/not found objects that expect an application.

@danielpeach @icfantv PTAL

### CODE REVIEW

File: app/scripts/modules/core/loadBalancer/loadBalancer.states.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        controllerProvider: ['$stateParams', 'cloudProviderRegistry',
+          ($stateParams: ILoadBalancerDetailsStateParams,
+           cloudProviderRegistry: CloudProviderRegistry) => {
+            return cloudProviderRegistry.getValue($stateParams.provider, 'loadBalancer.detailsController');
+        }],
+        controllerAs: 'ctrl'
+      }
+    },
+    resolve: {
+      loadBalancer: ['$stateParams', function($stateParams: ILoadBalancerDetailsStateParams) {
```
Comment (danielpeach) [inline app/scripts/modules/core/loadBalancer/loadBalancer.states.ts]: Would be more consistent to use an arrow function here.

File: app/scripts/modules/core/navigation/stateHelper.provider.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import {module, copy} from 'angular';
+import {INestedState} from './state.provider';
+
+export class StateHelper implements ng.IServiceProvider {
+
+  private registeredStates: string[] = [];
+
+  static get $inject() { return ['$stateProvider']; }
+
+  constructor(private $stateProvider: any) {}
```
Comment (danielpeach) [inline app/scripts/modules/core/navigation/stateHelper.provider.ts]: Any reason not to type `$stateProvider`?

Comment (anotherchrisberry) [inline app/scripts/modules/core/navigation/stateHelper.provider.ts]: No good reason - added type, thanks for catching

File: app/scripts/modules/core/loadBalancer/loadBalancer.states.ts (hunk 1 above)
Comment (anotherchrisberry) [inline app/scripts/modules/core/loadBalancer/loadBalancer.states.ts]: agreed, it is now an arrow function

File: app/scripts/modules/core/delivery/delivery.states.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+export interface IExecutionDetailsStateParams extends IApplicationStateParams {
+  executionId: string;
+  refId?: string;
+  stage?: string;
+  step?: string;
+  details?: string;
+}
+
+export const DELIVERY_STATES = 'spinnaker.core.deliver.states';
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/delivery.states.ts]: nice catch

File: app/scripts/modules/core/serverGroup/serverGroup.states.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        controllerProvider: ['$stateParams', 'cloudProviderRegistry',
+          ($stateParams: IServerGroupStateParams,
+           cloudProviderRegistry: CloudProviderRegistry) => {
+            return cloudProviderRegistry.getValue($stateParams.provider, 'serverGroup.detailsController');
+        }],
+        controllerAs: 'ctrl'
+      }
+    },
+    resolve: {
+      serverGroup: ['$stateParams', function($stateParams: IServerGroupStateParams) {
```
Comment (icfantv) [inline app/scripts/modules/core/serverGroup/serverGroup.states.ts]: arrow function?

File: app/scripts/modules/core/projects/projects.states.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        title: 'Dashboard'
+      }
+    },
+  };
+
+  const project: INestedState = {
+    name: 'project',
+    url: '/projects/{project}',
+    resolve: {
+      projectConfiguration: ['$stateParams', 'projectReader', function($stateParams: IProjectStateParms, projectReader: any) {
```
Comment (icfantv) [inline app/scripts/modules/core/projects/projects.states.ts]: arrow function?

File: app/scripts/modules/core/serverGroup/serverGroup.states.ts (hunk 4 above)
Comment (anotherchrisberry) [inline app/scripts/modules/core/serverGroup/serverGroup.states.ts]: yep, good catch

File: app/scripts/modules/core/projects/projects.states.ts (hunk 5 above)
Comment (anotherchrisberry) [inline app/scripts/modules/core/projects/projects.states.ts]: yep, good catch

**Comment by anotherchrisberry** [comment]
cc @zanthrash - hopefully this is easy enough to follow with your new FP changes

**Pull request merged**